DirectProgramming/DPC++FPGA/Tutorials/Features/mem_channels/src/CMakeLists.txt

```diff
@@ -0,0 +1,90 @@
+else()
+    message(STATUS "Configuring the design to run on FPGA board ${FPGA_BOARD}")
+endif()
+
+# This is a Windows-specific flag that enables exception handling in host code
+if(WIN32)
+    set(WIN_FLAG "/EHsc")
+endif()
+
+# A DPC++ ahead-of-time (AoT) compile processes the device code in two stages.
+# 1. The "compile" stage compiles the device code to an intermediate representation (SPIR-V).
+# 2. The "link" stage invokes the compiler's FPGA backend before linking.
+#    For this reason, FPGA backend flags must be passed as link flags in CMake.
+set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fintelfpga -DFPGA_EMULATOR")
+set(EMULATOR_LINK_FLAGS "-fintelfpga")
+set(HARDWARE_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fintelfpga")
+set(HARDWARE_LINK_FLAGS "-fintelfpga -Xshardware -Xsboard=${FPGA_BOARD} ${USER_HARDWARE_FLAGS}")
+# use cmake -D USER_HARDWARE_FLAGS=<flags> to set extra flags for FPGA backend compilation
+
+###############################################################################
+### FPGA Emulator
+###############################################################################
+# To compile in a single command:
+#    dpcpp -fintelfpga -DFPGA_EMULATOR mem_channels.cpp -o mem_channels.fpga_emu
+# CMake executes:
+#    [compile] dpcpp -fintelfpga -DFPGA_EMULATOR -o mem_channels.cpp.o -c mem_channels.cpp
+#    [link]    dpcpp -fintelfpga mem_channels.cpp.o -o mem_channels.fpga_emu
+add_executable(${EMULATOR_TARGET} ${SOURCE_FILE})
+set_target_properties(${EMULATOR_TARGET} PROPERTIES COMPILE_FLAGS "${EMULATOR_COMPILE_FLAGS}")
+set_target_properties(${EMULATOR_TARGET} PROPERTIES LINK_FLAGS "${EMULATOR_LINK_FLAGS}")
+add_custom_target(fpga_emu DEPENDS ${EMULATOR_TARGET})
+
+###############################################################################
+### Generate Report
+###############################################################################
+# To compile manually:
+#   dpcpp -fintelfpga -Xshardware -Xsboard=<FPGA_BOARD> -fsycl-link=early mem_channels.cpp -o mem_channels_report.a
+set(FPGA_EARLY_IMAGE_INTERLEAVING ${TARGET_NAME}_interleaving_report.a)
+set(FPGA_EARLY_IMAGE_NO_INTERLEAVING ${TARGET_NAME}_no_interleaving_report.a)
+# The compile output is not an executable, but an intermediate compilation result unique to DPC++.
+add_executable(${FPGA_EARLY_IMAGE_INTERLEAVING} ${SOURCE_FILE})
+add_executable(${FPGA_EARLY_IMAGE_NO_INTERLEAVING} ${SOURCE_FILE})
+add_custom_target(report DEPENDS ${FPGA_EARLY_IMAGE_INTERLEAVING} ${FPGA_EARLY_IMAGE_NO_INTERLEAVING})
+set_target_properties(${FPGA_EARLY_IMAGE_INTERLEAVING} PROPERTIES COMPILE_FLAGS "${HARDWARE_COMPILE_FLAGS}")
+set_target_properties(${FPGA_EARLY_IMAGE_INTERLEAVING} PROPERTIES LINK_FLAGS "${HARDWARE_LINK_FLAGS} -fsycl-link=early")
+
+if (FPGA_BOARD STREQUAL "intel_a10gx_pac:pac_a10")
+  set_target_properties(${FPGA_EARLY_IMAGE_NO_INTERLEAVING} PROPERTIES COMPILE_FLAGS "${HARDWARE_COMPILE_FLAGS} -DNO_INTERLEAVING")
```

Comment:
What happens when we compile this design to some non-PAC BSP?  You've assumed everything other than pac_a10 will have four channels.  What happens when we compile to our a10_ref or s10_ref BSPs, which only have one channel?  Will we get errors?  What about agilex_ref?  I'm not sure if it has 2 or 4 channels.

Reply:
Here's what I did to address this:
* Add  two macros: `TWO_CHANNELS` and `FOUR_CHANNELS`.
* Define `TWO_CHANNELS` for PAC A10.
* Define `FOUR_CHANNELS` for PAC S10.
* Not define anything for any other board but ask the reader to set this macro correctly for whatever customer board they're using. For
So, a10_ref, s10_ref, no-interleaving won't change the behavior because only 1 channels exists anyways. For Agilex, the no-interleaving version won't have any macros defined. Once we start mentioning Agilex in the tutorials, we then have to define the `FOUR_CHANNELS` macro for that target.